example/snake/reduce.hpp

```diff
@@ -0,0 +1,143 @@
+#include <cstdlib>
+
+namespace sn {
+
+namespace detail {
+
+template <typename T>
+T wrap(T val, T min, T max)
+{
+    if (val < min)
+        return max - (min - val) + 1;
+    if (val > max)
+        return min + (val - max) - 1;
+    return val;
+}
+
+inline direction left(direction initial)
+{
+    if (initial == direction::up || initial == direction::down)
+        return direction::left;
+    return initial;
+}
+
+inline direction right(direction initial)
+{
+    if (initial == direction::up || initial == direction::down)
+        return direction::right;
+    return initial;
+}
+
+inline direction up(direction initial)
+{
+    if (initial == direction::left || initial == direction::right)
+        return direction::up;
+    return initial;
+}
+
+inline direction down(direction initial)
+{
+    if (initial == direction::left || initial == direction::right)
+        return direction::down;
+    return initial;
+}
+
+inline point_t move_forward(point_t pos, direction dir)
+{
+    switch (dir) {
+    case direction::left:
+        return {x(pos) - 1, y(pos)};
+    case direction::up:
+        return {x(pos), y(pos) - 1};
+    case direction::right:
+        return {x(pos) + 1, y(pos)};
+    case direction::down:
+        return {x(pos), y(pos) + 1};
+    }
+
+    BOOST_UNREACHABLE_RETURN(pos);
+}
+
+inline snake_t::body_t move_forward(snake_t::body_t body, direction dir)
+{
+    const auto head = move_forward(body.front(), dir);
+    std::rotate(body.begin(), body.end() - 1, body.end());
+    body[0] = head;
+    return body;
+}
+
+inline bool in_bounds(point_t p)
+{
+    return x(p) >= 0 && x(p) < game_t::width && y(p) >= 0 && y(p) < game_t::height;
+}
+
+} // namespace detail
+
+inline game_t reduce(game_t game, action_t action)
+{
+    return boost::apply_visitor(
+        lager::visitor{
+            [&](go_left) {
+                game.snake.dir = detail::left(game.snake.dir);
+                return game;
+            },
+            [&](go_right) {
+                game.snake.dir = detail::right(game.snake.dir);
+                return game;
+            },
+            [&](go_up) {
+                game.snake.dir = detail::up(game.snake.dir);
+                return game;
+            },
+            [&](go_down) {
+                game.snake.dir = detail::down(game.snake.dir);
+                return game;
+            },
+            [&](tick) {
+                const auto prevBack = game.snake.body.back();
+                auto body =
+                    detail::move_forward(game.snake.body, game.snake.dir);
+                const auto head = body.front();
+
+                if (std::any_of(body.begin() + 1,
+                                body.end(),
+                                [&](const auto& p) { return p == head; }) ||
+                    !detail::in_bounds(head)) {
+                    game.over = true;
+                    return game;
+                }
+
+                if (prevBack == game.apple_pos) {
```

Comment:
`prevBack` vs `prev_back`? ;)

Reply:
I like switching it up every now and then... keeps the reader on their toes!